### PR TITLE
HPCC-27170 Fix remote index filter/limit issues

### PR DESCRIPTION
**Pull request by jakesmith**
Fix a number of issues related to reading indexes remotely
(via dafilesrv streaming), where filtering or limits are present.

The indexread/count/normalize activities optionally pass a threshold
limit to the remote side, to suppress excessive unecessary rows
being serialized back.
This is total remoteLimit disabled if there are is any post filtering,
because it is not known how many remote rows will eventually be
needed.
Three issues fixed:
1) IndexCount was using helper->hasMatchFilter() but hasFilter()
was set instead. This was causing it to send a remoteLimit, and
therefore not receive rows back which would have matched the post-
filter.
2) IndexNormalize was also basing remoteLimit on hasMatchFilter().
However, it neither sets hasFilter or hasMatchFilter. Change to
assume no remoteLimit in indexnormalize case.
3) getRowLimit was being used as a remoteLimit boundary, meaning
it would falsely hit the limit, when post filters would reduce
it below that limit.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
> <Warning><Code>4523</Code><Filename>indexlimit3.ecl</Filename><Line>55</Line><Source>eclcc</Source><Message>Neither LIMIT() nor CHOOSEN() supplied for index read on  ~keyx</Message></Warning>

@ghalliday - for some reason I can't reproduce this locally (in Release or Debug).
and also not sure why it's saying that the index read need a LIMIT or CHOOSEN ?

**Comment by jakesmith** [comment]
Also wondering if this should target 7.12.x

**Comment by ghalliday** [comment]
@richardkchapman note the persist failure due to the new roxie message....
@jakesmith since this changes the behaviour of existing queries it should really target master.  @richardkchapman thoughts?

**Comment by jakesmith** [comment]
>@richardkchapman note the persist failure due to the new roxie message....

I'm blind, assume it was the thor run of the query that was failing.
Ok, well, I can suppress the warning I guess in the test.

> @jakesmith since this changes the behaviour of existing queries it should really target master. @richardkchapman thoughts?

True, it does change the behaviour. It's a very subtle bug though, which to the users is non-deterministic.
It depends whether their copy of the index is local or remote (+ whether the limit effects the overall results or not)

**Comment by jakesmith** [comment]
@ghalliday - please review carefully.

**Comment by jakesmith** [comment]
New test in roxie is hitting:
<Exception><Source>Roxie</Source><Message>LimitSkipException (in Index Normalize 27)</Message></Exception>

(it didn't in 8.6.x)

Not sure why at the moment.
@richardkchapman - any initial thoughts?

**Review by ghalliday — APPROVED** [review]
@richardkchapman I think this is good/safe.  Even if there was a problem it would only have the effect of removing an optimization.

**Comment by richardkchapman** [comment]
@jakesmith Probably because of 

    HPCC-21884 Index Normalize ignoring LIMIT SKIP and generating exception

(fixed in 8.0)

**Comment by jakesmith** [comment]
> @jakesmith Probably because of
> 
> ```
> HPCC-21884 Index Normalize ignoring LIMIT SKIP and generating exception
> ```
> 
> (fixed in 8.0)

@richardkchapman - ok, could remove this particular test in 7.12 - to avoid keep seeing it on future 7.12 point builds, and then issue new PR to add it back in 8.0.x ?

**Comment by richardkchapman** [comment]
@jakesmith That sounds sensible

**Review by richardkchapman — APPROVED** [review]

File: testing/regress/ecl/indexlimit3.ecl [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+// indexcount with limits. NB: if exceeded, count will be 1
+ OUTPUT(COUNT(LIMIT(fi, 3, ONFAIL(TRANSFORM(RECORDOF(fi), SELF.info := '**ROW LIMIT HIT IN INDEXCOUNT**'; SELF := [])))));
+ OUTPUT(COUNT(LIMIT(fi, 5, KEYED, ONFAIL(TRANSFORM(RECORDOF(fi), SELF.info := '**KEYED LIMIT HIT IN INDEXCOUNT**'; SELF := [])))));
+ OUTPUT(EXISTS(fi));
+
+// indexread with choosen
+ OUTPUT(CHOOSEN(fi, 1));
+
+// indexnormalize
+// OUTPUT(LIMIT(fi.kkeys(ckey>=992), 3, ONFAIL(TRANSFORM(RECORDOF(crec), SELF.info := '**ROW LIMIT HIT IN INDEXNORMALIZE**'; SELF := []))));
```
Comment (richardkchapman) [inline testing/regress/ecl/indexlimit3.ecl:92]: This one should not be commented out

File: testing/regress/ecl/indexlimit3.ecl [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+// indexread with LIMIT
+ OUTPUT(LIMIT(fi, 2, ONFAIL(TRANSFORM(RECORDOF(fi), SELF.info := '**ROW LIMIT HIT IN INDEXREAD**'; SELF := []))));
+ OUTPUT(LIMIT(fi, 4, KEYED, ONFAIL(TRANSFORM(RECORDOF(fi), SELF.info := '**KEYED LIMIT HIT IN INDEXREAD**'; SELF := []))));
+
+// indexcount with limits. NB: if exceeded, count will be 1
+ OUTPUT(COUNT(LIMIT(fi, 2, ONFAIL(TRANSFORM(RECORDOF(fi), SELF.info := '**ROW LIMIT HIT IN INDEXCOUNT**'; SELF := [])))));
+ OUTPUT(COUNT(LIMIT(fi, 4, KEYED, ONFAIL(TRANSFORM(RECORDOF(fi), SELF.info := '**KEYED LIMIT HIT IN INDEXCOUNT**'; SELF := [])))));
+
+// indexnormalize
+ OUTPUT(LIMIT(fi.kkeys(ckey>=992), 2, ONFAIL(TRANSFORM(RECORDOF(crec), SELF.info := '**ROW LIMIT HIT IN INDEXNORMALIZE**'; SELF := []))));
```
Comment (richardkchapman) [inline testing/regress/ecl/indexlimit3.ecl:75]: This is the one that will fail on roxie

**Review by richardkchapman — CHANGES_REQUESTED** [review]

**Review by richardkchapman — APPROVED** [review]

**Comment by jakesmith** [comment]
@richardkchapman - github test passed now (apart from usual helm chart failure)

**Pull request merged**